File: shop-backend/src/@core/libs/access-control-list/services/permission-container.service.ts
import LoggerService from '@core/libs/logger/logger.system'
import RedisService from '@core/libs/redis/redis.service'
import PermissionRepository from '../repositories/permission.repository'
import { PermissionModel } from '../models/permission.model'
import { permissions } from '../const/permission'
import {
	OnApplicationBootstrap,
	OnModuleInit
} from '@core/enums/app-lifecycle.enum'

export interface PermissionInterface {
	name: string
	slug: string
	tag: string[]
	fromModule?: string
	module?: string
}

export class PermissionContainerService
	implements OnModuleInit, OnApplicationBootstrap
{
	constructor(
		private readonly _loggerService: LoggerService,
		private readonly _redisService: RedisService,
		private readonly _permissionRepository: PermissionRepository
	) {}

	static container: PermissionInterface[] = []
	static containerSlug: string[] = []

	static getPermissions(module: string) {
		return PermissionContainerService.container.filter(
			(permission) => permission.module === module
		)
	}

	static register(permissions: PermissionInterface[], module?: string) {
		for (const permission of permissions) {
			if (!this.containerSlug.includes(permission.slug)) {
				this.container.push({
					...permission,
					module
				})
				this.containerSlug.push(permission.slug)
			}
		}
	}

	onModuleInit(): any {
		PermissionContainerService.register(permissions, 'core')
	}

	onApplicationBootstrap(): any {
		setTimeout(() => {
			this.seedNewPermissions()
		}, 4000)
	}

	private seedNewPermissions() {
		this._permissionRepository
			.findAll()
			.then((dbPermissions: PermissionModel[]) => {
				const dbPermissionSlugs = dbPermissions.map(
					(permission) => permission.slug
				)
				const permissionEntities: PermissionInterface[] =
					PermissionContainerService.container
						.filter((permission) => {
							return !dbPermissionSlugs.includes(permission.slug)
						})
						.map((permission) => permission)

				if (permissionEntities.length) {
					this._permissionRepository
						.create(permissionEntities)
						.then((result) => console.log(result))
						.catch((error) => {
							this._loggerService.error(
								'Save new Permissions error',
								error
							)
						})
				}

				//
				dbPermissions.filter((permission) => {
					return !PermissionContainerService.containerSlug.includes(
						permission.slug
					)
				})
			})
	}
}

File: shop-admin-backend/src/main/java/com/demo/admin/core/accessControlList/services/PermissionContainerService.java
package com.demo.admin.core.accessControlList.services;
import jakarta.annotation.PostConstruct;
import lombok.AllArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.boot.context.event.ApplicationReadyEvent;

import com.demo.admin.core.accessControlList.dtos.CreatePermissionDto;
import com.demo.admin.core.accessControlList.models.PermissionModel;
import com.demo.admin.core.accessControlList.repositories.PermissionRepository;
import com.demo.admin.core.accessControlList.constants.PermissionCoreConstants;
import org.springframework.stereotype.Service;


import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Collectors;


@AllArgsConstructor
@Service
public class PermissionContainerService {

    private final PermissionRepository _permissionRepository;

    /* ================= STATIC CONTAINER ================= */

    private static final List<CreatePermissionDto> container = new CopyOnWriteArrayList<>();
    private static final Set<String> containerSlug = new HashSet<>();

    public static List<CreatePermissionDto> getPermissions(String module) {
        return container.stream()
                .filter(p -> Objects.equals(p.getModule(), module))
                .toList();
    }

    public static synchronized void register(
            List<CreatePermissionDto> permissions,
            String module
    ) {
        for (CreatePermissionDto permission : permissions) {
            if (!containerSlug.contains(permission.getSlug())) {
                permission.setModule(module);
                container.add(permission);
                containerSlug.add(permission.getSlug());
            }
        }
    }

    @PostConstruct
    public void onModuleInit() {
        register(PermissionCoreConstants.PERMISSIONS, "core");
    }


    @EventListener(ApplicationReadyEvent.class)
    public void onApplicationBootstrap() {
        new Timer().schedule(
            new TimerTask() {
                @Override
                public void run() {
                    seedNewPermissions();
                }
            },
            4000
        );
    }

    private void seedNewPermissions() {
        try {
            List<PermissionModel> dbPermissions = _permissionRepository.findAll();

            Set<String> dbPermissionSlugs = dbPermissions.stream()
                    .map(PermissionModel::getSlug)
                    .collect(Collectors.toSet());

            List<CreatePermissionDto> newPermissions = container.stream()
                    .filter(p -> !dbPermissionSlugs.contains(p.getSlug()))
                    .toList();

            if (!newPermissions.isEmpty()) {
                _permissionRepository.saveAll(
                        newPermissions.stream()
                                .map(PermissionModel::fromDto)
                                .toList()
                );
            }

        } catch (Exception e) {
//            log.error("Save new Permissions error", e);
        }
    }

}
